app/components/UI/Ramp/hooks/useBalance.ts

```diff
@@ -26,51 +33,67 @@ const defaultReturn = {
 };
 
 interface Asset {
-  address: string;
+  address?: string;
   decimals: number;
+  chainId?: string;
+  assetId?: string;
 }
 
 export default function useBalance(asset?: Asset) {
   const accountsByChainId = useSelector(selectAccountsByChainId);
+  ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+  const multiChainTokenBalance = useSelector(selectMultichainBalances);
+  ///: END:ONLY_INCLUDE_IF
   const chainId = useSelector(selectEvmChainId);
   const selectedAddress = useSelector(
     selectSelectedInternalAccountFormattedAddress,
   );
+  const selectedAccount = useSelector(selectSelectedInternalAccount);
   const conversionRate = useSelector(selectConversionRate);
   const currentCurrency = useSelector(selectCurrentCurrency);
   const tokenExchangeRates = useSelector(selectContractExchangeRates);
   const balances = useSelector(selectContractBalances);
 
-  if (!asset) {
-    return defaultReturn;
-  }
-
-  const assetAddress = safeToChecksumAddress(asset.address);
-
-  if (!assetAddress) {
+  if (
+    !asset ||
+    (!asset.address && !asset.assetId) ||
+    !selectedAddress ||
+    !selectedAccount
+  ) {
     return defaultReturn;
   }
 
   let balance, balanceFiat, balanceBN;
-  if (assetAddress === NATIVE_ADDRESS) {
+
+  ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+  if (asset.assetId) {
+    // CAIP19 asset identifier
+    const assetCaip19Identifier = `${asset.chainId}/${asset.assetId}`;
+    const assetBalance =
+      multiChainTokenBalance?.[selectedAccount.id]?.[assetCaip19Identifier];
+    balance = `${assetBalance?.amount ?? ''} ${
+      assetBalance?.unit ?? ''
+    }`.trim();
+  } else
+  ///: END:ONLY_INCLUDE_IF
+  if (asset.address === NATIVE_ADDRESS) {
```

Comment:
the formatter hates this and will set it to be like this if we are not careful 💀 
```suggestion
  }
  ///: END:ONLY_INCLUDE_IF
  else if (asset.address === NATIVE_ADDRESS) {
```

Reply:
would something like this work better?

```ts
export default function useBalance(asset?: Asset) {
  const accountsByChainId = useSelector(selectAccountsByChainId);
  ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
  const multiChainTokenBalance = useSelector(selectMultichainBalances);
  ///: END:ONLY_INCLUDE_IF
  const chainId = useSelector(selectEvmChainId);
  const selectedAddress = useSelector(
    selectSelectedInternalAccountFormattedAddress,
  );
  const selectedAccount = useSelector(selectSelectedInternalAccount);
  const conversionRate = useSelector(selectConversionRate);
  const currentCurrency = useSelector(selectCurrentCurrency);
  const tokenExchangeRates = useSelector(selectContractExchangeRates);
  const balances = useSelector(selectContractBalances);

  if (
    !asset ||
    (!asset.address && !asset.assetId) ||
    !selectedAddress ||
    !selectedAccount
  ) {
    return defaultReturn;
  }

  let balance, balanceFiat, balanceBN;

  ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
  if (asset.assetId) {
    // CAIP19 asset identifier
    const assetCaip19Identifier = `${asset.chainId}/${asset.assetId}`;
    const assetBalance =
      multiChainTokenBalance?.[selectedAccount.id]?.[assetCaip19Identifier];

    balance = `${assetBalance?.amount ?? ''} ${assetBalance?.unit ?? ''}`.trim();
    // If fenced code is removed, we simply skip this block and move on.
  }
  ///: END:ONLY_INCLUDE_IF

  if (!balance && asset.address === NATIVE_ADDRESS) {
    const chainKey = toHexadecimal(chainId);
    if (!accountsByChainId[chainKey]) {
      return defaultReturn;
    }

    balance = renderFromWei(accountsByChainId[chainKey][selectedAddress]?.balance);
    balanceBN = hexToBN(accountsByChainId[chainKey][selectedAddress]?.balance);
    balanceFiat = weiToFiat(balanceBN, conversionRate, currentCurrency);

  } else if (!balance && asset.address) {
    const assetAddress = safeToChecksumAddress(asset.address);
    const exchangeRate = tokenExchangeRates?.[assetAddress as Hex]?.price;

    balance =
      assetAddress && assetAddress in balances
        ? renderFromTokenMinimalUnit(balances[assetAddress], asset.decimals ?? 18)
        : 0;

    balanceFiat = balanceToFiat(balance, conversionRate, exchangeRate, currentCurrency);

    balanceBN =
      assetAddress && assetAddress in balances
        ? hexToBN(balances[assetAddress])
        : null;
  }

  return { balance, balanceFiat, balanceBN };
}
```

Reply:
Done in [b2a0cb6](https://github.com/MetaMask/metamask-mobile/pull/13948/commits/b2a0cb6e1fe1feeb4b215bd7a02b7f950d05c807)